### DISCUSSION TITLE
Pass mesh and time information to the DAObjFunc class

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#549] [General]
Hi,

This is more of a design of the code type of question.

I am trying to perform interpolation between two meshes on the fly in DAFoam. I have managed to search the OpenFOAM source code and retrieve the relevant functions that perform this task. I have even written a small application in native OpenFOAM to do this task for me.

Now I want to do the same with DAFoam but I am facing so many issues. Primarily because of the structure. I would like to do my interpolation at the end of all the primal solver calls (because it is costly) so I can have a separate solver and remove the printAllObjFuncs() function so that the functioncall to derived calcObjFunc() does not take place at every primal solver iteration (happens only at the end). But in order to do my interpolation, I need to create a new mesh object (and I want to have this option for only some objective function child classes) which requires the time object as well. I have tried everything to declare this in the DAObjFunc class, tried a lot of tricks but I have not been successful.

Is there a way to translate something like this 
```
const Foam::word polyMeshPath = "target";
const Foam::IOobject meshIO("target", polyMeshPath, runTime, Foam::IOobject::MUST_READ);

// Create the fvMesh object using the updated path to the polyMesh directory
Foam::fvMesh target(meshIO);
```

into an equivalent code in DAFoam in the DAObjFunc parent class?

I even tried having multiple constructors since I come to understand that there is a runTime election table which picks the child class at runtime based on the type information but I was not successful. Any hints on where to begin would be highly useful!

**Comment by friedenhe** [comment]
What error did you get when you put the above code in DAObjFunc? Regarding your question, I suggest you implement your new objFunc as a function (e.g., called myNewObj)in DASover.C first. Then, at the end of the DASimpleFOam loop, you call this->myNewObj(). If this approach works, you can then move it to the DAObjFunc class.

**Reply by uttamcadambi07** [reply]
To answer your question - I added these lines in DAObjFunc.C
```
    const IOobject meshIO("target",polyMeshPath, runTime,IOobject::MUST_READ);

    // Create the fvMesh object using the updated path to the polyMesh directory
    fvMesh target(meshIO);
```
and I added these lines in DAObjFunc.H
```
    const word polyMeshPath = "target";
    
    const fvMesh target;
```

I get this error
```
DAObjFunc/DAObjFunc.C: In constructor ‘Foam::DAObjFunc::DAObjFunc(const Foam::fvMesh&, const Foam::DAOption&, const Foam::DAModel&, const Foam::DAIndex&, const Foam::DAResidual&, Foam::word, Foam::word, const Foam::dictionary&)’:
DAObjFunc/DAObjFunc.C:41:48: error: no matching function for call to ‘Foam::fvMesh::fvMesh()’
   41 |       daField_(mesh, daOption, daModel, daIndex)
      |                                                ^
In file included from /home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvCFD.H:7,
                 from DAObjFunc/DAObjFunc.H:16,
                 from DAObjFunc/DAObjFunc.C:8:
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:204:9: note: candidate: ‘Foam::fvMesh::fvMesh(const Foam::IOobject&, Foam::pointField&&, Foam::faceList&&, Foam::cellList&&, bool)’
  204 |         fvMesh
      |         ^~~~~~
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:204:9: note:   candidate expects 5 arguments, 0 provided
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:192:9: note: candidate: ‘Foam::fvMesh::fvMesh(const Foam::IOobject&, Foam::pointField&&, Foam::faceList&&, Foam::labelList&&, Foam::labelList&&, bool)’
  192 |         fvMesh
      |         ^~~~~~
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:192:9: note:   candidate expects 6 arguments, 0 provided
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:188:9: note: candidate: ‘Foam::fvMesh::fvMesh(const Foam::IOobject&, Foam::zero, bool)’
  188 |         fvMesh(const IOobject& io, const zero, bool syncPar=true);
      |         ^~~~~~
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:188:9: note:   candidate expects 3 arguments, 0 provided
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:184:18: note: candidate: ‘Foam::fvMesh::fvMesh(const Foam::IOobject&)’
  184 |         explicit fvMesh(const IOobject& io);
      |                  ^~~~~~
/home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/finiteVolume/lnInclude/fvMesh.H:184:18: note:   candidate expects 1 argument, 0 provided
DAObjFunc/DAObjFunc.C:62:50: error: ‘runTime’ was not declared in this scope
   62 |     const IOobject meshIO("target",polyMeshPath, runTime,IOobject::MUST_READ);
      |                                                  ^~~~~~~

```

So it says runTime is not declared in this scope. So traditionally, runTime object of Time class is created using createTime.H. But in DAFoam, I noticed it is done differently using setArgs.H, setRootCasePython.H, createTimePython.H, createMeshPython.H. These have modified OpenFOAM codes slightly. I also noticed that DASolver.H has the autoPtr<Time> runTimePtr_; smart pointer created which is then reset using the new keyword.

So far, this has been my understanding and progress. So the goal is to try to initialize the mesh (for which the runTime object is required). So we need to find a way to either create or pass the runTimePtr to DAObjFunc.

**Reply by uttamcadambi07** [reply]
Okay then! brainstorming this idea with you - I have tried it and it compiles (yet to check if it works).

So I created a static smart pointer ```static autoPtr<Time> runTime``` in the DAObjFunc class and then assign the runTimePtr_ from DASolver class to it. This way I don't have to instantiate the DAObjFunc class to be able to assign the runTime pointer and don't have to create a new Time object. So now with this, I used another smart pointer ```autoPtr<fvMesh> target``` and dereferenced the smart pointer to the static Time object to create my fvMesh object for the second mesh in DAObjFunc class. 

This is a bit of a workaround but I think it checks all boxes.

**Reply by uttamcadambi07** [reply]
Update on this: I tested it and I seem to be getting this error:
```
--> FOAM FATAL ERROR: 
[1] object of type N4Foam4TimeE is unallocated
[1] 
[1]     From function T& Foam::autoPtr<T>::operator*() [with T = Foam::Time]
[1]     in file /home/uttamcadambi/dafoam_v2/OpenFOAM/OpenFOAM-v1812/src/OpenFOAM/lnInclude/autoPtrI.H at line 214.

```

And this is after the statement 
```
    Info << "Initializing mesh and runtime for DASolver" << endl;

```
in DASolver.C executes. It is after this statement that I initialize the static Time object pointer in DAObjFunc class with that of runTimePtr_(). Perhaps I am making a mistake in the manner in which it is being initialised? Clearly this error means that the pointer is trying to dereference but it does not hold any value 

I do the initializsing this way
```
    DAObjFunc::runTime_ = runTimePtr_();

```

and the static pointer  in DAObjFunc class is 
```
    static autoPtr<Time> runTime_;
```

**Comment by friedenhe** [comment]
> To answer your question - I added these lines in DAObjFunc.C
> 
> ```
>     const IOobject meshIO("target",polyMeshPath, runTime,IOobject::MUST_READ);
> 
>     // Create the fvMesh object using the updated path to the polyMesh directory
>     fvMesh target(meshIO);
> ```

Can you change the above `runTime` to `runTimePtr_()` and recompile? I don't see a reason why you need to initialize the runTime object again. It is already initialized and it is called runTimePtr_, you can get the object by using `runTimePtr_()`.

**Reply by uttamcadambi07** [reply]
Oh even if it is in DASolver class? The problem I was facing is to access this runtime pointer from the DAObjFunc class because that’s where I need it

**Comment by friedenhe** [comment]
If runTimePtr_() is not available, you can use `mesh_.time()` instead. They are the same. There is no need to create a separate runTime object.

**Reply by uttamcadambi07** [reply]
Hi. Forgot to mention that yes this works for me. But again, facing a new issue. 
I will provide some more details in a few hours to explain the problem a bit better, thanks for the help!
